// arb-ts/dist/lib/abi/factories/IChallenge__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IChallenge__factory = void 0;
const ethers_1 = require("ethers");
class IChallenge__factory {
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IChallenge__factory = IChallenge__factory;
const _abi = [
    {
        inputs: [],
        name: 'asserter',
        outputs: [
            {
                internalType: 'address',
                name: '',
                type: 'address',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'challenger',
        outputs: [
            {
                internalType: 'address',
                name: '',
                type: 'address',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'clearChallenge',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [],
        name: 'currentResponderTimeLeft',
        outputs: [
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'contract IOneStepProof[]',
                name: '_executors',
                type: 'address[]',
            },
            {
                internalType: 'address',
                name: '_resultReceiver',
                type: 'address',
            },
            {
                internalType: 'bytes32',
                name: '_executionHash',
                type: 'bytes32',
            },
            {
                internalType: 'uint256',
                name: '_maxMessageCount',
                type: 'uint256',
            },
            {
                internalType: 'address',
                name: '_asserter',
                type: 'address',
            },
            {
                internalType: 'address',
                name: '_challenger',
                type: 'address',
            },
            {
                internalType: 'uint256',
                name: '_asserterTimeLeft',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: '_challengerTimeLeft',
                type: 'uint256',
            },
            {
                internalType: 'contract ISequencerInbox',
                name: '_sequencerBridge',
                type: 'address',
            },
            {
                internalType: 'contract IBridge',
                name: '_delayedBridge',
                type: 'address',
            },
        ],
        name: 'initializeChallenge',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [],
        name: 'lastMoveBlock',
        outputs: [
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'timeout',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
];
