// arb-ts/dist/lib/abi/factories/IERC677Receiver__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IERC677Receiver__factory = void 0;
const ethers_1 = require("ethers");
class IERC677Receiver__factory {
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IERC677Receiver__factory = IERC677Receiver__factory;
const _abi = [
    {
        inputs: [
            {
                internalType: 'address',
                name: '_sender',
                type: 'address',
            },
            {
                internalType: 'uint256',
                name: '_value',
                type: 'uint256',
            },
            {
                internalType: 'bytes',
                name: '_data',
                type: 'bytes',
            },
        ],
        name: 'onTokenTransfer',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
];
