// arb-ts/dist/lib/abi/factories/IExitLiquidityProvider__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IExitLiquidityProvider__factory = void 0;
const ethers_1 = require("ethers");
class IExitLiquidityProvider__factory {
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IExitLiquidityProvider__factory = IExitLiquidityProvider__factory;
const _abi = [
    {
        inputs: [
            {
                internalType: 'address',
                name: 'dest',
                type: 'address',
            },
            {
                internalType: 'address',
                name: 'erc20',
                type: 'address',
            },
            {
                internalType: 'uint256',
                name: 'amount',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: 'exitNum',
                type: 'uint256',
            },
            {
                internalType: 'bytes',
                name: 'liquidityProof',
                type: 'bytes',
            },
        ],
        name: 'requestLiquidity',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
];
