// arb-ts/dist/lib/abi/factories/IMessageProvider__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IMessageProvider__factory = void 0;
const ethers_1 = require("ethers");
class IMessageProvider__factory {
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IMessageProvider__factory = IMessageProvider__factory;
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: 'uint256',
                name: 'messageNum',
                type: 'uint256',
            },
            {
                indexed: false,
                internalType: 'bytes',
                name: 'data',
                type: 'bytes',
            },
        ],
        name: 'InboxMessageDelivered',
        type: 'event',
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: 'uint256',
                name: 'messageNum',
                type: 'uint256',
            },
        ],
        name: 'InboxMessageDeliveredFromOrigin',
        type: 'event',
    },
];
