// arb-ts/dist/lib/abi/factories/INodeFactory__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.INodeFactory__factory = void 0;
const ethers_1 = require("ethers");
class INodeFactory__factory {
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.INodeFactory__factory = INodeFactory__factory;
const _abi = [
    {
        inputs: [
            {
                internalType: 'bytes32',
                name: '_stateHash',
                type: 'bytes32',
            },
            {
                internalType: 'bytes32',
                name: '_challengeHash',
                type: 'bytes32',
            },
            {
                internalType: 'bytes32',
                name: '_confirmData',
                type: 'bytes32',
            },
            {
                internalType: 'uint256',
                name: '_prev',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: '_deadlineBlock',
                type: 'uint256',
            },
        ],
        name: 'createNode',
        outputs: [
            {
                internalType: 'address',
                name: '',
                type: 'address',
            },
        ],
        stateMutability: 'nonpayable',
        type: 'function',
    },
];
