// arb-ts/dist/lib/abi/factories/L1Buddy__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.L1Buddy__factory = void 0;
const ethers_1 = require("ethers");
class L1Buddy__factory {
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.L1Buddy__factory = L1Buddy__factory;
const _abi = [
    {
        inputs: [
            {
                internalType: 'address',
                name: '_inbox',
                type: 'address',
            },
            {
                internalType: 'address',
                name: '_l2Deployer',
                type: 'address',
            },
        ],
        stateMutability: 'nonpayable',
        type: 'constructor',
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: 'uint256',
                name: 'seqNum',
                type: 'uint256',
            },
            {
                indexed: false,
                internalType: 'address',
                name: 'l2Address',
                type: 'address',
            },
        ],
        name: 'DeployBuddyContract',
        type: 'event',
    },
    {
        inputs: [],
        name: 'codeHash',
        outputs: [
            {
                internalType: 'bytes32',
                name: '',
                type: 'bytes32',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'bool',
                name: 'success',
                type: 'bool',
            },
        ],
        name: 'finalizeBuddyDeploy',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [],
        name: 'inbox',
        outputs: [
            {
                internalType: 'contract IInbox',
                name: '',
                type: 'address',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'uint256',
                name: 'maxSubmissionCost',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: 'maxGas',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: 'gasPriceBid',
                type: 'uint256',
            },
            {
                internalType: 'bytes',
                name: 'contractInitCode',
                type: 'bytes',
            },
        ],
        name: 'initiateBuddyDeploy',
        outputs: [
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256',
            },
        ],
        stateMutability: 'payable',
        type: 'function',
    },
    {
        inputs: [],
        name: 'l2Buddy',
        outputs: [
            {
                internalType: 'address',
                name: '',
                type: 'address',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'l2Connection',
        outputs: [
            {
                internalType: 'enum L1Buddy.L2Connection',
                name: '',
                type: 'uint8',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'l2Deployer',
        outputs: [
            {
                internalType: 'contract BuddyDeployer',
                name: '',
                type: 'address',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
];
