// arb-ts/dist/lib/abi/factories/RetryableTicketCreator__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RetryableTicketCreator__factory = void 0;
const ethers_1 = require("ethers");
class RetryableTicketCreator__factory {
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.RetryableTicketCreator__factory = RetryableTicketCreator__factory;
const _abi = [
    {
        inputs: [
            {
                internalType: 'address',
                name: 'destAddr',
                type: 'address',
            },
            {
                internalType: 'uint256',
                name: 'l2CallValue',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: 'maxSubmissionCost',
                type: 'uint256',
            },
            {
                internalType: 'address',
                name: 'excessFeeRefundAddress',
                type: 'address',
            },
            {
                internalType: 'address',
                name: 'callValueRefundAddress',
                type: 'address',
            },
            {
                internalType: 'uint256',
                name: 'maxGas',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: 'gasPriceBid',
                type: 'uint256',
            },
            {
                internalType: 'bytes',
                name: 'data',
                type: 'bytes',
            },
        ],
        name: 'createRetryableTicket',
        outputs: [],
        stateMutability: 'payable',
        type: 'function',
    },
];
